Add a dashboard pager component (#1168)

djjuhasz opened the pull request:
Create a general Vue.js Pager component that can be used for all dashboard lists.

- Create a Vue.js Pager component
- Use the Pager component on the `ingest/sips` and `storage/aips` pages
- Update the page URL when the page changes with the current page
  number, to allow bookmarking a specific page in the results
- Remove pager properties and methods from the AIP and SIP stores

djjuhasz committed:
Add a dashboard pager component (#1168)

Create a general Vue.js Pager component that can be used for all
dashboard lists.

- Create a Vue.js Pager component
- Use the Pager component on the `ingest/sips` and `storage/aips` pages
- Update the page URL when the page changes with the current page
  number, to allow bookmarking a specific page in the results
- Remove pager properties and methods from the AIP and SIP stores
- Create a ResultCounter component and include it on the AIPs and SIPs
  pages to report number of found and shown records

[skip-codecov]

jraddaoui commented on dashboard/src/components/Pager.vue:
```diff
@@ -0,0 +1,137 @@
+
+      <!-- First page and ellipses-->
+      <template v-if="visiblePages[0] > 1">
+        <li class="page-item" :class="{ disabled: currentPage === 1 }">
```
I think this will always be enabled, right? Or would it be possible to have the current page outside of the visible pages?

djjuhasz replied:
Yes, good point. :)

jraddaoui commented on dashboard/src/components/Pager.vue:
```diff
@@ -0,0 +1,137 @@
+            >1</a
+          >
+        </li>
+        <li class="d-none d-sm-block" aria-hidden="true">
```
Should we check if the first visible page is 2 and not show this item?

jraddaoui commented on dashboard/src/components/Pager.vue:
```diff
@@ -0,0 +1,137 @@
+
+      <!-- Last page and ellipses-->
+      <template v-if="visiblePages[visiblePages.length - 1] < totalPages">
+        <li class="d-none d-sm-block" aria-hidden="true">
```
Similar to the comment above but at the end. If `visiblePages[visiblePages.length - 1] === totalPages - 1`, I wouldn't show the ellipsis.

jraddaoui commented on dashboard/src/components/Pager.vue:
```diff
@@ -0,0 +1,137 @@
+        <li class="d-none d-sm-block" aria-hidden="true">
+          <a href="#" class="page-link disabled">…</a>
+        </li>
+        <li class="page-item" :class="{ disabled: currentPage === totalPages }">
```
Same, I think this will always be enabled if `totalPages`  is outside `visiblePages`.

jraddaoui reviewed (changes requested):
Looks great to me, thanks @djjuhasz!

It would be nice to remove the existing pager from the AIPs browse page and reuse this one too, I think that should happen in this PR for consistency. Doing so we may realize there is still some duplication around, I wonder if it would be a good idea to pass the fetch function directly to the pager, instead of emitting and doing it outside, but that can happen in a following PR if we think it deserves the effort. Same with the pager related getters from the stores, I think we could move the entire logic into the pager component.

djjuhasz committed:
Add AIP generation to sample data generator

- Add a CLI argument to Go package generator script to select AIP or
  SIP generation
- Update the load.sh script to generate and load sample AIP data as
  well as the sample SIPs

djjuhasz commented:
@jraddaoui I've made the changes you suggested and switched to the Pager component for paging the `storage/aips` page.  This is ready for another review, thanks.

jraddaoui commented on dashboard/src/components/Pager.vue:
```diff
@@ -0,0 +1,156 @@
+});
+
+const totalPages = computed(() => {
+  // Calculate total pages based on offset and limit
```
Based on total and limit.

jraddaoui commented on dashboard/src/pages/ingest/sips/index.vue:
```diff
@@ -204,7 +223,7 @@ watch(
     <div class="text-muted mb-3">
       Showing {{ sipStore.page.offset + 1 }} -
-      {{ sipStore.lastResultOnPage }} of
+      {{ sipStore.page.offset + sipStore.page.limit }} of
```
This may not be true in the final page, where the amount of records doesn't have to match the limit. Also, it currently shows `1 - 0 of 0` if there are no results, I think now it will show `1 to 20 of 0` in that case.

djjuhasz replied:
I've added a `ResultCounter` component to show a more readable count of the shown results on the SIPs and AIPs pages.  It now shows:

- "No results found" (0 results)
- "Found 1 result" (1 result)
- "Found 10 results" (more than 1 result, but less than two pages of results)
- "Showing 1 - 20 of 35 results" (two or more pages of results)

djjuhasz replied:
I've made the `ResultCounter` it's own commit for review, but I think I'll squash it with the the dashboard pager commit before merging.

jraddaoui reviewed (changes requested):
Thanks @djjuhasz! It would be nice to make those counts a bit more solid.

jraddaoui reviewed (approved):
Thanks @djjuhasz, really nice!
